components/clm/src/biogeochem/LSparseMatMod.F90

```diff
@@ -0,0 +1,428 @@
+    integer , pointer :: icol(:)     !col id for each val
+    integer , pointer :: pB(:)       !first nonzero column id (as in val) in each row
+    integer , pointer :: ncol(:)     !number of nonzero columns in eacho row
+    integer :: szcol                 ! number of columns of the matrix
+    integer :: szrow
+  contains
+    procedure, public :: init
+  end type sparseMat_type
+  real(r8), parameter :: tiny_val=1.e-14_r8
+  public :: spm_list_init, spm_list_insert, spm_list_to_mat
+  public :: flux_correction
+contains
+
+!---------------------------------------------------------------
+  function create_spm_type()
+    !
+    ! ! DESCRIPTION:
+    !
+    ! create an object of type sparseMat_type.
+    ! Right now it is purposely empty
+   class(sparseMat_type), pointer :: spm
+   class(sparseMat_type), pointer ::  create_spm_type
+   allocate(spm)
+   create_spm_type => spm
+  end function create_spm_type
+
+!---------------------------------------------------------------
+  subroutine init(this,nelms,nrows,szcol)
+  !
+  !DESCRIPTION
+  ! initialize the sparse matrix
+  implicit none
+  class(sparseMat_type), intent(inout) :: this
+  integer, intent(in) :: nelms
+  integer, intent(in) :: nrows
+  integer, intent(in) :: szcol
+
+  allocate(this%val(nelms)) ;  this%val(:)  = 0._r8
+  allocate(this%icol(nelms));  this%icol(:) = 0
+  allocate(this%pB(nrows));    this%pB(:) = 0
+  allocate(this%ncol(nrows));  this%ncol(:) = 0
+  this%szcol = szcol
+  this%szrow= nrows
+  end subroutine init
+
+!---------------------------------------------------------------
+  subroutine spm_pack(Amat, nelms)
+
+  implicit none
+  real(r8), dimension(:,:),intent(in) :: Amat
+  integer, intent(in) :: nelms
+
+  class(sparseMat_type), pointer :: spm
+  integer :: nrows, szcol
+  integer :: ii,jj,kk
+  logical :: first
+  nrows=size(Amat,1)
+  szcol=size(Amat,2)
+  allocate(spm, source=create_spm_type())
+
+  call spm%init(nelms, nrows, szcol)
+
+  kk = 0
+  do ii = 1, nrows
+    first =.true.
+    do jj = 1, szcol
+      if(abs(amat(ii,jj))>tiny_val)then
+        kk = kk + 1
+        spm%val(kk) = amat(ii,jj)
+        spm%icol(kk)= jj
+        if(first)then
+          spm%pB(ii) = jj
+        endif
+        spm%ncol(ii)=spm%ncol(ii)+1
+        first=.false.
+      endif
+    enddo
+  enddo
+
+  end subroutine spm_pack
+
+!---------------------------------------------------------------
+
+  subroutine spm_unpack(spm, bmat)
+
+  implicit none
+  class(sparseMat_type), intent(in) :: spm
+  real(r8), pointer :: bmat(:,:)
+
+  integer :: nrs, ncs
+  integer :: i, j, id
+
+  nrs = size(spm%ncol)
+  ncs = spm%szcol
+  allocate(bmat(nrs,ncs))
+
+  do i = 1, nrs
+    do j = 1, spm%ncol(i)
+      id=j+spm%pB(i)
+      bmat(i,spm%icol(id))=spm%val(id)
+    enddo
+  enddo
+
+  end subroutine spm_unpack
+
+!---------------------------------------------------------------
+
+  subroutine spm_axpy(nx, ny, a, x, spm, y, errinfo)
```

Comment:
Again, is the function of `spm_axpy` obvious to everyone but me? 😆  A one-line comment explaining the purpose of this would be useful.